### PR TITLE
Adding events "meetingReconnected" signalingDropped" "receivingAudioDropped"  to eventDidReceive

### PR DESCRIPTION
**Pull request by harshitOberoi**
**Issue #:**
N/A

**Description of changes:**
Added events meetingReconnected, signalingDropped, and receivingAudioDropped to `eventDidReceive` where as currently these events are pushed to meeting history.

**Testing**
Added 3 new Tests:
1. publish event for Signalling Dropped when Web Socket Failed
2. publish event for Signalling Dropped when Web Socket Closed
3. publish event for Signalling Dropped when Web Socket error

Updated Test:
notifies the connection is poor

1. Have you successfully run `npm run build:release` locally? 
Yes
2. How did you test these changes? 
    a) Run Browser Demo App.
    b) Connect to a meeting
    c) Use Network Link Conditioner to mimic 100% loss.
    d) Open console section in developers tab  and filter by "Received an event"
    e) we should receive "receivingAudioDropped" and  "signalingDropped". event.
    f) Enable wifi and we should receive "meetingReconnected" event.
    g) Check gamma ingestion logs and we should receive all the three events.

3. Can these changes be tested using the demo application? If yes, which demo application can be used to test it?
Yes, Browser Demo 

4. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved?
No

5. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved?
No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by nainkunal933** [comment]
@harshitOberoi Do you need the package-lock.json file in your PR? Usually we do not want to include lock files if we have do not have any changes in package.json.

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,15 +14,18 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 - The Amazon Voice Focus support check, `VoiceFocusDeviceTransformer.isSupported`, now warns to the logger when run in an iframe, and can be configured to fail in that case.
+- Add events `meetingReconnected`, `signalingDropped` and `receivingAudioDropped` to `eventDidReceive` by publishing them as stand alone events where as these events are currently being pushed into meeting History. 
 
 ### Changed
 
 - Clarify why not use default downlink policy with simulcast.
 - Corrected argument `isUnifiedPlan` in `withBandwidthRestriction` to `isFirefox`. Also marked as deprecated since we no longer use it.
 - Update data message limit in the API Overview guide.
 
+
```
[the remark this answers is not in the document]
Comment (harshitOberoi) [reply CHANGELOG.md]: removed

File: CHANGELOG.md [hunk 2]
```diff
@@ -13,6 +13,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added Android WebView Sample UI test to workflow.
 - Add a new optional API `getVideoTileForAttendeeId` in `VideoTileController` and raise the `tileWillBePausedByDownlinkPolicy` event for empty video tiles.
 - Amazon Voice Focus will now trigger the `onCPUWarning` method on the `VoiceFocusTransformDeviceDelegate` when using the default inline execution mode. This will be called when the browser fails to schedule the worklet in a timely fashion (_e.g._, when the meeting code is running in an iframe /subframe) or when changes in the CPU environment (_e.g._, thermal throttling) cause the worklet to take too long for each audio render quantum.
+- The Amazon Voice Focus support check, `VoiceFocusDeviceTransformer.isSupported`, now warns to the logger when run in an iframe, and can be configured to fail in that case.
```
Comment (devalevenkatesh) [inline CHANGELOG.md:16]: The "The Amazon Voice Focus support check.." changelog is already existing below in the changelog so you might have mistakenly added it here.

File: CHANGELOG.md [hunk 3]
```diff
@@ -13,6 +13,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- Add events `meetingReconnected`, `signalingDropped` and `receivingAudioDropped` to `eventDidReceive` by publishing them as stand alone events where as these events are currently being pushed into meeting History.
```
Comment (devalevenkatesh) [inline CHANGELOG.md:17]: Should we break this into two sentences something like below?
"Add `meetingReconnected`, `signalingDropped` and `receivingAudioDropped` meeting events to the meeting events published using `eventDidReceive` `AudioVideoObserver`. Currently, these meeting events were only included in the meeting history attribute when a meeting event is published."

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 4]
```diff
@@ -1534,6 +1534,9 @@ export class DemoMeetingApp
       case 'videoInputSelected':
       case 'audioInputUnselected':
       case 'videoInputUnselected':
+      case 'meetingReconnected':
+      case 'receivingAudioDropped':
+      case 'signalingDropped':
       case 'attendeePresenceReceived': {
         // Exclude the "meetingHistory" attribute for successful events.
```
Comment (devalevenkatesh) [inline demos/browser/app/meetingV2/meetingV2.ts:1541]: This is not part of your change but since we are adding lifecycle events here, you can change this comment with "successful" -> "published" may be.

File: demos/browser/package-lock.json [hunk 5]
```diff
@@ -20504,4 +20504,4 @@
       "dev": true
     }
   }
-}
```
Comment (devalevenkatesh) [inline demos/browser/package-lock.json:20507]: You can revert this change as well, and hence this file won't show in the PR.

**Review by simmkyu — CHANGES_REQUESTED** [review]
You can add unit tests for the `meetingReconnected` event using the existing `DefaultAudioVideoController` test. For example, in the following test, we add an observer to receive events and ensure the order of four events.

https://github.com/aws/amazon-chime-sdk-js/blob/c0c78983bdd6637a922c68237ffc0206eba37028/test/audiovideocontroller/DefaultAudioVideoController.test.ts#L2478-L2484

⠀

For the `meetingReconnected` event, you can do the same:

https://github.com/aws/amazon-chime-sdk-js/blob/c0c78983bdd6637a922c68237ffc0206eba37028/test/audiovideocontroller/DefaultAudioVideoController.test.ts#L1718

File: test/task/MonitorTask.test.ts [hunk 6]
```diff
@@ -958,6 +958,45 @@ describe('MonitorTask', () => {
       expect(spy.called).to.be.false;
     });
 
+    it('publish event for Signalling Dropped when Web Socket Failed', () => {
```
Comment (nainkunal933) [inline test/task/MonitorTask.test.ts:961]: Can you add tests for `meetingReconnected` event as well? Look into `DefaultAudioVideoController.test.ts` for some existing reconnect tests.

**Review by nainkunal933 — CHANGES_REQUESTED** [review]

File: test/task/MonitorTask.test.ts (hunk 6 above)
Comment (nainkunal933) [reply test/task/MonitorTask.test.ts:961]: `DefaultAudioVideoController.test.ts` might be a better candidate for this test. Also, looking at it my request and @simmkyu's request is quite identical. So, I will resolve my comment so you can just focus on Kyu's comment.

File: src/task/MonitorTask.ts [hunk 7]
```diff
@@ -286,7 +286,7 @@ export default class MonitorTask
       this.logger.info(`unusable audio warning is now: ${unusableAudioWarningValue}`);
       if (unusableAudioWarningValue === 0) {
         this.context.poorConnectionCount += 1;
-        this.context.eventController?.pushMeetingState('receivingAudioDropped');
+        this.context.eventController?.publishEvent('receivingAudioDropped');
```
[the remark this answers is not in the document]
Comment (harshitOberoi) [reply src/task/MonitorTask.ts]: @simmkyu  As publishEvent expects attibutes of type `AudioVideoEventAttributes` i made changes to send the complete arguments.

[the remark this answers is not in the document]
Comment (harshitOberoi) [reply src/task/MonitorTask.ts]: Added the arguments

File: src/audiovideocontroller/DefaultAudioVideoController.ts [hunk 8]
```diff
@@ -1016,7 +1016,7 @@ export default class DefaultAudioVideoController
       this.sessionStateController.perform(SessionStateControllerAction.FinishConnecting, () => {
         /* istanbul ignore else */
         if (this.eventController) {
-          this.eventController.pushMeetingState('meetingReconnected');
+          this.eventController.publishEvent('meetingReconnected');
```
[the remark this answers is not in the document]
Comment (harshitOberoi) [reply src/audiovideocontroller/DefaultAudioVideoController.ts]: Added the arguments as well as added events in meeting attributes

File: guides/06_Meeting_Events.md [hunk 9]
```diff
@@ -99,6 +99,9 @@ The Chime SDK for JavaScript sends these meeting events.
 |`videoInputSelected`    |The camera was selected.
 |`videoInputUnselected`  |The camera was removed. You called `meetingSession.audioVideo.chooseVideoInputDevice` with `null`.
 |`videoInputFailed`      |The camera selection failed.
+|`signalingDropped`      |The WebSocket failed or closed with an error.
+|`meetingReconnected`    |The meeting reconnected.
```
[the remark this answers is not in the document]
Comment (harshitOberoi) [reply guides/06_Meeting_Events.md]: Fixed

File: src/eventcontroller/EventName.ts [hunk 10]
```diff
@@ -13,6 +13,9 @@ type EventName =
   | 'audioInputFailed'
   | 'videoInputSelected'
   | 'videoInputUnselected'
-  | 'videoInputFailed';
+  | 'videoInputFailed'
+  | 'meetingReconnected'
```
[the remark this answers is not in the document]
Comment (harshitOberoi) [reply src/eventcontroller/EventName.ts]: fixed

File: CHANGELOG.md (hunk 2 above)
Comment (harshitOberoi) [reply CHANGELOG.md]: @devalevenkatesh fixed

File: src/task/MonitorTask.ts [hunk 11]
```diff
@@ -345,7 +363,7 @@ export default class MonitorTask
       event.type === SignalingClientEventType.WebSocketFailed
     ) {
       if (!this.hasSignalingError) {
-        this.context.eventController?.pushMeetingState('signalingDropped');
+        this.context.eventController?.publishEvent('signalingDropped');
```
Comment (simmkyu) [inline src/task/MonitorTask.ts:366]: Sorry, I should've mentioned in the previous revision. You can also include attributes for the `signalingDropped` event. You may create a private helper method to generate the same attributes for `signalingDropped` and `receivingAudioDropped`.

**Review by simmkyu — COMMENTED** [review]
I left one comment for adding attributes to the `signalingDropped` event. Other than that, LGTM!

File: src/task/MonitorTask.ts (hunk 11 above)
Comment (harshitOberoi) [reply src/task/MonitorTask.ts]: Updated Signalling dropped event to contain audioVieo Attributes

**Review by nainkunal933 — APPROVED** [review]

File: guides/06_Meeting_Events.md [hunk 12]
```diff
@@ -131,16 +134,16 @@ The following table describes attributes for a meeting.
 
 |Attribute|Description|Included in
 |--|--|--
-|`attendeePresenceDurationMs`|The time taken for the attendee to be present in the meeting.<br><br>Unit: Milliseconds|`attendeePresenceReceived`, `meetingEnded`, `meetingFailed`
-|`iceGatheringDurationMs`|The time taken for connection's ICE gathering state to complete.<br><br>Unit: Milliseconds|`meetingStartSucceeded`, `meetingStartFailed`, `meetingEnded`, `meetingFailed`
-|`maxVideoTileCount`|The maximum number of simultaneous video tiles shared during the meeting. This includes a local tile (your video), remote tiles, and content shares.<br><br>Unit: Count|`meetingStartSucceeded`, `meetingStartFailed`, `meetingEnded`, `meetingFailed`
-|`meetingDurationMs`|The time that elapsed between the beginning (`AudioVideoObserver.audioVideoDidStart`) and the end (`AudioVideoObserver.audioVideoDidStop`) of the meeting.<br><br>Unit: Milliseconds|`meetingEnded`, `meetingFailed`
+|`attendeePresenceDurationMs`|The time taken for the attendee to be present in the meeting.<br><br>Unit: Milliseconds|`attendeePresenceReceived`, `meetingEnded`, `meetingFailed`, `meetingReconnected`, `receivingAudioDropped`
```
Comment (simmkyu) [inline guides/06_Meeting_Events.md:137]: Please update the doc with the `signalingDropped` event.

Comment (harshitOberoi) [reply guides/06_Meeting_Events.md]: fixed

File: guides/06_Meeting_Events.md [hunk 13]
```diff
@@ -131,16 +134,16 @@ The following table describes attributes for a meeting.
...
-|`meetingDurationMs`|The time that elapsed between the beginning (`AudioVideoObserver.audioVideoDidStart`) and the end (`AudioVideoObserver.audioVideoDidStop`) of the meeting.<br><br>Unit: Milliseconds|`meetingEnded`, `meetingFailed`
+|`attendeePresenceDurationMs`|The time taken for the attendee to be present in the meeting.<br><br>Unit: Milliseconds|`attendeePresenceReceived`, `meetingEnded`, `meetingFailed`, `meetingReconnected`, `receivingAudioDropped`, `signalingDropped`
+|`iceGatheringDurationMs`|The time taken for connection's ICE gathering state to complete.<br><br>Unit: Milliseconds|`meetingStartSucceeded`, `meetingStartFailed`, `meetingEnded`, `meetingFailed`, `meetingReconnected`, `receivingAudioDropped`, `signalingDropped`
+|`maxVideoTileCount`|The maximum number of simultaneous video tiles shared during the meeting. This includes a local tile (your video), remote tiles, and content shares.<br><br>Unit: Count|`meetingStartSucceeded`, `meetingStartFailed`, `meetingEnded`, `meetingFailed`, `meetingReconnected`, `receivingAudioDropped`
```
Comment (simmkyu) [inline guides/06_Meeting_Events.md:139]: The `maxVideoTileCount` row misses the `signalingDropped` event. Other than that, the PR looks good to me!

Comment (harshitOberoi) [reply guides/06_Meeting_Events.md]: Added.

File: CHANGELOG.md [hunk 14]
```diff
@@ -5,7 +5,16 @@ All notable changes to this project will be documented in this file.
...
 
-## [Unreleased]
+##[Unreleased]
+
+### Added
+- Add events `meetingReconnected`, `signalingDropped` and `receivingAudioDropped` to `eventDidReceive` by publishing them as stand alone events. Currently, these events were only included in the meeting history attribute when a meeting event is published. 
+
+
+### Changed
+- Move `toLowerCasePropertyNames` inside `Utils.ts` and add test coverage.
```
Comment (simmkyu) [inline CHANGELOG.md:15]: Thanks for moving this item into the **Unreleased/Changed** section, but I don't see the item under the **[2.17.0] - 2021-09-08/Changed** section. Could you check you rebased the PR correctly?
https://github.com/aws/amazon-chime-sdk-js/blame/master/CHANGELOG.md#L30

**Review by simmkyu — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 14 above)
Comment (harshitOberoi) [reply CHANGELOG.md:15]: Kunal merged his changelog md changes by mistake so asked me to remove his comment from [2.17.0] to unreleased.

Comment (simmkyu) [reply CHANGELOG.md:15]: Sorry for the confusion. If you have removed it, the GitHub diff UI should show that.

image

The PR is still out of sync. Just rebase again.
> This branch is out-of-date with the base branch

**Review by simmkyu — APPROVED** [review]
